### ISSUE TITLE
How to use target network in replay

### ISSUE THREAD

**Issue by kk-55** [deepmind/dqn#8]
Hey guys!

In the following code snippet I wanna replicate the described utilization of a target network.
My network has a state as input and give me the Q-values for all the possible actions.
To get the target Q-value for my currently considered state-action pair I use this if-else clause.
But what is with the target Q-values for the "state-remaining-action pairs" (see both lines with `target_f = self.model.predict(state)` or `target_f = self.target_model.predict(state)`)? Do I use the "policy network" (`self.model`) or the "target network" (`self.target_model`)?
In the case of the "policy network" the losses for all the "state-remaining-action pairs" are equal to zero and just for the currently considered state-action it is different from zero (commented out). In the second case all the losses are different from zero. Right?

Or maybe both cases are reasonable, please give me some advice. Thanks in advance!

```
# replay the history and train the model
def replay(self, batch_size):

	# choose batch_size "unique" (i.e. repeats could be selected) random samples from memory of the agent
	minibatch = random.sample(self.memory, batch_size)

	# experience replay Q-learning (with a target network)
	for state, action, reward, next_state, premature_termination, all_done in minibatch:
		# case if episode terminates at next step
		if premature_termination or all_done:
			target = reward
		else:
			target = (reward + self.gamma *
	                          np.amax(self.target_model.predict(next_state)[0]))
		''' self.model or self.target_model '''
		#target_f = self.model.predict(state)
		target_f = self.target_model.predict(state)
		target_f[0][action] = target    # approximate target value(s)
		self.model.fit(state, target_f, epochs=1, verbose=0)

	# decrease epsilon
	if self.epsilon > self.epsilon_min:
		self.epsilon *= self.epsilon_decay
```

**Comment by GeorgOstrovski** [comment]
In regular Q-learning as implemented in DQN, indeed, the other actions do not get updated - corresponding to the commented-out variant in your code. One way of putting this is that the loss is only a function of the 'taken' action (`action` in your code example). If your code requires you to implement this in terms of targets rather than losses, than in fact, you can use the policy network's own outputs as targets for those actions. 

There are other algorithms that (in different contexts, in different ways) also update the other action values. E.g [this paper](https://arxiv.org/abs/1704.03732), where learning happens from demonstration trajectories, and in addition to learning the value of the demonstrated actions, other (presumably sub-optimal) actions are being 'pushed down' by a certain margin to ensure that the resulting policy will prefer the demonstrated actions over alternatives.
